File: 36.cpp
TreeNode* Convert(TreeNode *root){
    if(root == nullptr)
        return nullptr;
    TreeNode *p = root, *last = nullptr;
    ConvertNode(root, last);
    while(p -> left != nullptr)
        p = p -> left;
    return p;
}

void ConvertNode(TreeNode *curr, TreeNode *&lastNode){
    if(curr == nullptr)
        return;
    if(curr -> left != nullptr)
        ConvertNode(curr -> left, lastNode);
    if(lastNode != nullptr){
        lastNode -> right = curr;
        curr -> left = lastNode;
    }
    lastNode = curr;
    if(curr -> right != nullptr){
        ConvertNode(curr -> right, lastNode);
    }
}

File: 36.java
class Solution {
    // last 表示前一个结点
    private TreeNode last = null;
    public TreeNode Convert(TreeNode root) {
        if (root == null)
            return root;
        ConvertNode(root);
        while (root.left != null)
            root = root.left;
        return root;
    }
     
    private void ConvertNode(TreeNode root) {
        if (root == null)
            return;
        ConvertNode(root.left);
        root.left = last;
        if (last != null) {
            last.right = root;
        }
        last = root;
        ConvertNode(root.right);
    }
}
